rust/src/measurements/randomized_response/make_randomized_response.tex

```diff
@@ -0,0 +1,145 @@
+    the mechanism truthfully returns the same value \texttt{arg} with probability \texttt{prob},
+    otherwise it lies by selecting one of the other categories uniformly at random.
+    \item ...if \texttt{arg} is not in \texttt{categories}, 
+    it returns a category chosen uniformly at random.
+\end{itemize}
+
+\subsection*{PR History}
+\begin{itemize}
+    \item \vettingPR{490}
+\end{itemize}
+
+\section{Hoare Triple}
+
+\subsection*{Preconditions}
+\begin{itemize}
+    \item Variable \texttt{categories} must be a set with members of type \texttt{T}
+    \item Variable \texttt{prob} must be of type \texttt{QO}
+    \item Type \texttt{QO} must have trait \rustdoc{traits/trait}{Float}
+    \item The bit representation of type \texttt{QO} must support \texttt{ExactIntCast} to and from \texttt{usize}
+\end{itemize}
+
+\subsection*{Pseudocode}
+\lstinputlisting[language=Python,firstline=2,escapechar=|]{./pseudocode/make_randomized_response.py}
+
+\subsection*{Postcondition}
+
+\validMeasurement{\texttt{(categories, prob, T, QO)}}{\\ \texttt{make\_randomized\_response}}
+
+\section{Proof}
+
+\begin{proof} 
+\textbf{(Privacy guarantee.)} 
+    
+\begin{tcolorbox}
+    The proof assumes the following lemma.
+    \begin{lemma}
+        \texttt{sample\_uniform\_int\_below} and \texttt{sample\_bernoulli\_float} satisfy their postconditions.
+    \end{lemma}
+\end{tcolorbox}
+
+\texttt{sample\_uniform\_int\_below} and \texttt{sample\_bernoulli\_float} can only fail due to lack of system entropy. 
+This is usually related to the computer's physical environment and not the dataset. 
+The rest of this proof is conditioned on the assumption that these functions do not raise an exception. 
+
+Let $x$ and $x'$ be datasets that are \texttt{d\_in}-close with respect to \texttt{input\_metric}.
+Here, the metric is \texttt{DiscreteMetric} which enforces that $\din \geq 1$ if $x \ne x'$ and $\din = 0$ if $x = x'$. 
+If $x = x'$, then the output distributions on $x$ and $x'$ are identical, and therefore the max-divergence is 0.
+
+Now consider the case where $x \ne x'$. 
+For shorthand, we let $p$ represent \texttt{prob}, the probability of returning the input,
+and $t$ denote the number of categories.
+Note that all categories must be unique as the input data type is a set.
+This means duplicate categories cannot influence the output distribution.
+
+$t$ must be at least two, by pseudocode line \ref{line:num_cats}, as any fewer would not be useful.
+$p$ is restricted to $[1/t, 1.0)$ by pseudocode line \ref{line:range}, as any less would not be useful.
+
+We'll first consider all possible output probabilities, 
+and then use this to upper bound the ratio of any two probabilities.
+For any outcome $y \in \texttt{candidates}$, 
+the probability of observing $y$ is one of three values:
+
+\begin{enumerate}
+    \item When the mechanism is honest: 
+    \[
+        \Pr[M(x) = y | y = x] = p
+    \]
+
+    \item When the mechanism lies: 
```

Comment:
It's (1-p)/(t-1), since it's only considering one output at a time. This proof writes out the probabilities of observing any one given output under three different conditions. It does this because, to ensure privacy, we want to bound the change in probability for any one given output t.